src/gnutls_crypto.cpp

```diff
@@ -266,35 +267,39 @@ namespace oxen::quic
             throw std::runtime_error("gnutls_set_default_priority failed");
         }
 
-        if (creds.using_raw_pk)
+        if (alpns.size())
         {
-            // server accepts all ALPNs configured as "allowed", client sends singular
-            // ALPN configured as "outbound".
-            const gnutls_datum_t* proto{nullptr};
-            unsigned int alpn_count{0};
-            if (is_client)
-            {
-                proto = &creds.outbound_alpn;
-                alpn_count = 1;
-            }
-            else
+            std::vector<gnutls_datum_t> allowed_alpns;
+            for (auto& s : alpns)
             {
-                proto = &(creds.allowed_alpns[0]);
-                alpn_count = creds.allowed_alpns.size();
+                log::trace(log_cat, "GNUTLS adding \"{}\" to {} ALPNs", s, direction_string);
+                allowed_alpns.emplace_back(gnutls_datum_t{
+                        reinterpret_cast<uint8_t*>(const_cast<char*>(s.data())), static_cast<uint32_t>(s.size())});
             }
 
-            if (proto == nullptr or alpn_count == 0)
+            if (auto rv =
+                        gnutls_alpn_set_protocols(session, &(allowed_alpns[0]), allowed_alpns.size(), GNUTLS_ALPN_MANDATORY);
+                rv < 0)
             {
-                log::error(log_cat, "raw public key mode session init, but no ALPNs have been specified.");
-                throw std::runtime_error("raw public key mode session init, but no ALPN specified.");
+                log::error(log_cat, "gnutls_alpn_set_protocols failed: {}", gnutls_strerror(rv));
+                throw std::runtime_error("gnutls_alpn_set_protocols failed");
             }
-
-            if (auto rv = gnutls_alpn_set_protocols(session, proto, alpn_count, 0); rv < 0)
+        }
+        else  // set default, mandatory ALPN string
+        {
+            gnutls_datum_t alpn;
+            auto default_alpn = "default"s;
+            alpn.data = reinterpret_cast<uint8_t*>(default_alpn.data());
+            alpn.size = default_alpn.size();
```

Comment:
Minor optimization: You could have this hard-coded as a const member variable or something similar?

Reply:
yeah, fair point.  Will do.